Feat/create search bar

Vietprole opened the pull request:
# Preview: https://react-training-internship.vercel.app/
![image](https://github.com/user-attachments/assets/bb938f6c-82aa-40cb-b17d-2a7ad0548968)

Vietprole committed:
Create search bar UI

Vietprole committed:
Implement search function

Vietprole committed:
Write stories for search bar

xuanmai-agilityio commented on src/App.jsx:
```diff
@@ -42,24 +43,45 @@ function App() {
+    saveNoteToLocalStorage(notesAfterDelete);
+  }
+
+  const handleSearch = (searchPhrase) => {
```
This is not how we handle filter/search notes. The idea is to store the search phrase in another state and we compute the list of notes based on the **origin notes** and **search phrase**.
`const filteredNotes = notes.filter(note => note.content.includes(searchPhase));`
That means we always have a single source of truth.

Vietprole committed:
update .gitignore to ignore .vite

xuanmai-agilityio commented on src/App.jsx:
```diff
@@ -27,10 +29,9 @@ function getNotes(){
 
-  // Save notes to localStorage
-  useEffect(() => {
+  const saveNoteToLocalStorage = (notes) => {
```
The `useEffect` is a more appropriate way to handle sync data between state and local storage. Or you can refer to this article on the way to build a custom hook.
https://thomasderleth.de/keeping-react-state-and-local-storage-in-sync/

Vietprole committed:
Modify search logic

Using useEffect to sync data to localStorage
Use searchPhrase state to calculate the filtered notes

xuanmai-agilityio commented on src/App.jsx:
```diff
@@ -38,32 +46,46 @@ function App() {
-  }
+
+    const newNotes = [...notes, newNote];
+    setNotes(newNotes);
```
With note changes, we can use a callback function to handle state updates.
```setNotes(prevNotes => [...prevNotes, newNote])```

Check and update for other places.

Vietprole replied:
Ok, I've updated them

Vietprole committed:
Use callback function for setNotes

xuanmai-agilityio commented on src/App.jsx:
```diff
@@ -38,32 +46,43 @@ function App() {
   const handleDeleteNote = (id) => {
-    setNotes(notes.filter(note => note.id !== id));
-  }
+    setNotes(notes.filter((note) => note.id !== id));
```
This is not the way I mentioned. It should be like this
```
setNotes(prevNotes => prevNotes.filter((note) => note.id !== id));

xuanmai-agilityio replied:
Following this way, we don't need to refer to the `notes` value. Furthermore, when optimizing app performance, we can apply `useCallback` to this function with empty dependencies. So the delete button that receives this function as props will not re-render when `notes` changes.

Vietprole replied:
I understand, I will modify other setNotes to this format.

Vietprole committed:
Update setNotes to use previous notes instead
